src/ray/util/memory.cc

```diff
@@ -27,7 +29,10 @@ uint8_t *pointer_logical_and(const uint8_t *address, uintptr_t bits) {
 
 void parallel_memcopy(uint8_t *dst, const uint8_t *src, int64_t nbytes,
                       uintptr_t block_size, int num_threads) {
-  std::vector<std::thread> threadpool(num_threads);
+  static boost::asio::thread_pool *threadpool = NULL;
```

Comment:
`static auto *threadpool = new boost::asio::thread_pool(num_threads);`
Otherwise the initialization can race.
Also if we want to use `std::thread::hardware_concurrency()` we can make that a separate change instead?

Reply:
could it possible that this parameter is different for different calls?

Reply:
Good point. It does look like the `num_threads` parameter is constant [here](https://github.com/ray-project/ray/blob/b8dda0e3d09c8214db04b45852b07367615560fe/python/ray/includes/serialization.pxi#L5). But if `parallel_memcopy()` is called concurrently, using a static thread pool of `num_threads` may reduce parallelism. So the original approach of using `std::thread::hardware_concurrency()` may be necessary. In this case we can remove the `num_threads` parameter too.

Reply:
Do you mean that we always have `std::thread::hardware_concurrency()` concurrency for every call?

Reply:
I think the current patch is good in a sense that it still allows function users to specify the parallelism and it also caps the max parallelism to `hardware_concurrency()`.

Reply:
Ok, just fixing the static initialization would be good enough.

Reply:
what needs to be fix besides free the threadpool when ray exits?

Reply:
Sorry I wasn't clear. Freeing the pool is unnecessary for a static variable. But the `if (threadpool == NULL) {...}` block can have a race condition if it is called concurrently. We should rely on static variable initialization instead for thread safety (https://en.cppreference.com/w/cpp/language/storage_duration#Static_local_variables). Similar to the example [here](https://google.github.io/styleguide/cppguide.html#Static_and_Global_Variables), it can be
`static auto& threadpool = *new boost::asio::thread_pool(std::thread::hardware_concurrency())`.

Reply:
Also is posting to boost::asio::thread_pool thread safe?

Reply:
`static auto& threadpool = *new boost::asio::thread_pool(std::thread::hardware_concurrency())`
Would every caller create a new pool?

Reply:
The compiler generates an atomic operation to ensure the static variable is initialized only once even with concurrent callers. Only one thread pool will be created.

Reply:
Got it! Thanks. Being doing C for too long..

Reply:
Can you comment this 

> The compiler generates an atomic operation to ensure the static local variable is initialized only once even with concurrent callers. Only one thread pool will be created.

to the code?

Reply:
Also, if we keep hardware limit instead of passed num_threads, can we remove the parameter?

Reply:
Re `num_threads`: this was my confusion earlier too .. `num_threads` is also used to partition the memory region to be copied, so keeping it is closer to the existing behavior.